nighthawk/source/common/statistic_impl.cc

```diff
@@ -0,0 +1,216 @@
+  output.set_request_count(count());
+  output.mutable_mean()->set_nanos(mean());
+  output.mutable_stdev()->set_nanos(stdev());
+}
+
+StreamingStatistic::StreamingStatistic() : count_(0), mean_(0), sum_of_squares_(0) {}
+
+void StreamingStatistic::addValue(int64_t value) {
+  double delta, delta_n;
+  count_++;
+  delta = value - mean_;
+  delta_n = delta / count_;
+  mean_ += delta_n;
+  sum_of_squares_ += delta * delta_n * (count_ - 1);
+}
+
+uint64_t StreamingStatistic::count() const { return count_; }
+
+double StreamingStatistic::mean() const { return mean_; }
+
+double StreamingStatistic::variance() const { return sum_of_squares_ / (count_ - 1.0); }
+
+double StreamingStatistic::stdev() const { return sqrt(variance()); }
+
+std::unique_ptr<Statistic> StreamingStatistic::combine(const Statistic& statistic) {
+  const StreamingStatistic& a = *this;
+  const StreamingStatistic& b = dynamic_cast<const StreamingStatistic&>(statistic);
+  auto combined = std::make_unique<StreamingStatistic>();
+
+  combined->count_ = a.count() + b.count();
+  combined->mean_ = ((a.count() * a.mean()) + (b.count() * b.mean())) / combined->count_;
+  combined->sum_of_squares_ =
+      a.sum_of_squares_ + b.sum_of_squares_ +
+      pow(a.mean() - b.mean(), 2) * a.count() * b.count() / combined->count();
+  return combined;
+}
+
+InMemoryStatistic::InMemoryStatistic() : streaming_stats_(std::make_unique<StreamingStatistic>()) {}
+
+void InMemoryStatistic::addValue(int64_t sample_value) {
+  samples_.push_back(sample_value);
+  streaming_stats_->addValue(sample_value);
+}
+
+uint64_t InMemoryStatistic::count() const {
+  ASSERT(streaming_stats_->count() == samples_.size());
+  return streaming_stats_->count();
+}
+double InMemoryStatistic::mean() const { return streaming_stats_->mean(); }
+double InMemoryStatistic::variance() const { return streaming_stats_->variance(); }
+double InMemoryStatistic::stdev() const { return streaming_stats_->stdev(); }
+
+std::unique_ptr<Statistic> InMemoryStatistic::combine(const Statistic& statistic) {
+  auto combined = std::make_unique<InMemoryStatistic>();
+  const InMemoryStatistic& b = dynamic_cast<const InMemoryStatistic&>(statistic);
+
+  combined->samples_.insert(combined->samples_.end(), this->samples_.begin(), this->samples_.end());
+  combined->samples_.insert(combined->samples_.end(), b.samples_.begin(), b.samples_.end());
+  combined->streaming_stats_ = this->streaming_stats_->combine(*b.streaming_stats_);
+  return combined;
+}
+
+const int HdrStatistic::SIGNIFICANT_DIGITS = 4;
+
+HdrStatistic::HdrStatistic() : histogram_(nullptr) {
+  // Upper bound of 60 seconds (tracking in nanoseconds).
+  const uint64_t max_latency = 1000L * 1000 * 1000 * 60;
+
+  int status = hdr_init(1 /* min trackable value */, max_latency, HdrStatistic::SIGNIFICANT_DIGITS,
+                        &histogram_);
+  if (status != 0) {
+    ENVOY_LOG(error, "Failed to initialize HdrHistogram.");
+    histogram_ = nullptr;
```

Comment:
I wonder if this should be a `throw` and we crash out? Seems like an unlikely scenario.